Reorganize readme content and add ability to publish as GitHub pages.

intelkevinputnam opened the pull request:
1. .rst documents added as glue to support Sphinx output
2. conf.json, conf.py, requirements.txt, substitutions.txt are Sphinx support files.
3. Makefile - added html rule. make.bat is for Sphinx Windows support.
4. bad_refs.txt and fix_refs.py are needed as this is an atypical Sphinx deployment and some file dependencies need to be fixed after the build. (fix_refs.py called after sphinx build in Makefile).
5. .gitignore changed to exclude some local files I was using.
6. Multiple .md files added as part of reorganization.
7. publish.yml is github workflow - needs to be updated to work with intel/pmem-csi repo.

intelkevinputnam commented:
Forgot to mention processMDFiles.py, which I used for readme discovery. Could be removed without impact.

pohly commented:
Before I look into details, some high-level questions and suggestions:
- Is it possible to see or check the result of document changes *before* merging them? For individual .md files we could look at how GitHub rendered the modified file and while this was not always done and thus Markdown errors occasionally got past PR review, at least the option is there.
- How can we handle different docs for the different supported branches? We probably want to make available docs for master and the latest stable release (starting with the upcoming 0.7.0 - we won't backport this to 0.6.x).
- Can we reduce the top-level README.md to something that just provides a short explanation of the project and then links to the "docs" folder for details and the rendered documentation? Then people coming to the repo know where to continued without immediately being overwhelmed by the actual documentation. Worse, they might miss that there is a docs site...

intelkevinputnam commented:
@pohly 

* There are two recommended options for previewing locally. 
   1. Install reDocs GUI (Intel internal tool) and click "build" and "preview"
   2. "pip install -r requirements.txt" to install dependencies and run "make html" - build output will be in _build/html directory.
* How should it be organized when viewed? GitHub actions are pretty flexible. Were you thinking: intel.github.io/pmem-csi/latest, intel.github.io/pmem-csi/0.7.0, and intel.github.io/pmem-csi/0.8.0 as possible URLs?
* Yep. Shall I propose something?

pohly commented:
> "pip install -r requirements.txt" to install dependencies and run "make html" - build output will be in _build/html directory.

Can we automate this so that "make html" will run inside a container which (ideally) already has the necessary tools installed? We assume that the developer has the rights to invoke "docker run" and has Go installed, but try to limit the set of additional dependencies. If there's no such container image, then a Python venv inside _build would also be okay. The latter approach then makes Python and venv an additional requirement.

> How should it be organized when viewed? GitHub actions are pretty flexible. Were you thinking: intel.github.io/pmem-csi/latest, intel.github.io/pmem-csi/0.7.0, and intel.github.io/pmem-csi/0.8.0 as possible URLs?

We produce new patch releases automatically and frequently, usually without changing the content. "intel.github.io/pmem-csi/latest" and "intel.github.io/pmem-csi/0.7" (or in general, "intel.github.io/pmem-csi/x.y" for a "release-x.y" branch) sounds good to me.

> Yep. Shall I propose something?

Please do :smile:

intelkevinputnam commented:
> Can we automate this so that "make html" will run inside a container which (ideally) already has the necessary tools installed? 

I built up a Dockerfile a while ago, which I can modify for this purpose. Do you have a preference for base image (Ubuntu, Alpine, etc.)? My assumption was that the user would build the docs with something like: 

``` bash
sudo docker  run --rm  -v `pwd`:/home/sphinx -w /home/sphinx -i -t  <image name> make html
```

Any tips for using docker on IT managed Windows? I kept having access to local file system broken. 

> We produce new patch releases automatically and frequently, usually without changing the content. "intel.github.io/pmem-csi/latest" and "intel.github.io/pmem-csi/0.7" (or in general, "intel.github.io/pmem-csi/x.y" for a "release-x.y" branch) sounds good to me.
> 

So make the documentation HTML output one of the release artifacts, which gets deployed when the website is built.

> > Yep. Shall I propose something?
> 
> Please do 😄

Will do.

pohly commented:
I ran some experiments in our Jenkins worker and on my local machine. We don't have "venv" installed at the moment under Jenkins, but `apt-get install python3-venv` completes reasonably fast. Doing it without venv is worse because `apt-get install python3-pip` also downloads binutils and gcc, something that we don't need. Then `pip install` completes in a few seconds.

This is faster than downloading a (large!) Docker image like ddidier/sphinx-doc and we get full control over the content via the `requirements.txt`. It might also work under Windows, although I am not sure because there will be some POSIX shell dependencies, like `. _work/venv/bin/activate` to enter the venv.

Let me see whether I can come up with a patch for `make html` which uses venv...

intelkevinputnam commented:
Would you consider Tox? I realize it has more dependencies than just venv, but the project is already configured to use it.

pohly commented:
> Would you consider Tox? I realize it has more dependencies than just venv, but the project is already configured to use it.

What is Tox, and which project?

pohly commented:
> Let me see whether I can come up with a patch for make html which uses venv...

That seems to be working, see https://github.com/intel/pmem-csi/blob/d03cf74be569b333e274c2aaee5f732c0b49d45d/test/e2e/deploy/deploy.go#L387

intelkevinputnam commented:
> > Would you consider Tox? I realize it has more dependencies than just venv, but the project is already configured to use it.
> 
> What is Tox, and which project?

[Tox](https://pypi.org/project/tox/) is a Python test framework that uses venv under the hood. It does all the heavy lifting for you. I built a GUI app that uses it to encapsulate all the dependencies for building Sphinx projects. All that is needed are the tox.ini and setup.py already checked into this branch.

pohly commented:
> Is it possible to see or check the result of document changes before merging them?

It is: https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/view/change-requests/job/PR-571/Doc_20Site/

The job itself has a "Doc Site" item on the left: https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/view/change-requests/job/PR-571/

That's with the additional commit here: https://github.com/intel/pmem-csi/pull/571/commits/b59c9f3720e6d91a9d93b20736eb65e3deec0861

intelkevinputnam commented:
> > Is it possible to see or check the result of document changes before merging them?
> 
> It is: https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/view/change-requests/job/PR-571/Doc_20Site/
> 
> The job itself has a "Doc Site" item on the left: https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/view/change-requests/job/PR-571/
> 
> That's with the additional commit here: [b59c9f3](https://github.com/intel/pmem-csi/commit/b59c9f3720e6d91a9d93b20736eb65e3deec0861)

Nice. Couple of questions:

1. Is this infrastructure (azure) unique to K8 projects or can anyone use it?
2. How long do the artifacts of the PR build persist?

okartau commented:
> 2\. How long do the artifacts of the PR build persist?

there is no automatic cleanup configured, so in theory, for long time. But in practice, the system will run out of space, so I have made some manual deletions of older (multi-month) PR jobs.

pohly commented:
> 1. Is this infrastructure (azure) unique to K8 projects or can anyone use it?

Azure is shared with other teams, but the Jenkins instance is specific to the Cloud Native projects.

pohly commented:
> Tox is a Python test framework that uses venv under the hood. It does all the heavy lifting for you. I built a GUI app that uses it to encapsulate all the dependencies for building Sphinx projects.

Running venv isn't hard. I'm not sure whether Tox really has practical advantages for us. I'd rather keep everything as simple as possible.

> All that is needed are the tox.ini and setup.py already checked into this branch.

I haven't gotten to looking at individual files in this PR, so it's good that you mention those now - that way I don't need to ask :grin: 

We can't take those into PMEM-CSI even if we wanted to because the source URL is not public and might even be considered confidential (reveals information about the Intranet).

intelkevinputnam commented:
@pohly sounds good. Give me a couple of days.

intelkevinputnam commented on .gitignore:
```diff
@@ -1,3 +1,7 @@
 /_work
+/.tox
+Manifest
+/_build
```
OK. I'll change the Sphinx BUILDDIR to _output/html.

intelkevinputnam commented:
@pohly I've addressed quite a bit of your feedback. You can take a look at the latest deployed build here: https://intelkevinputnam.github.io/pmem-csi/. 

Branch and tag relative links are fairly easy. If it is really necessary to use a specific commit, I think we can solve the linking issue for commits using GitHub workflows to feed the value into the build.

pohly commented:
The structure looks good now. I noticed some dead links, for example in https://intelkevinputnam.github.io/pmem-csi/latest/install.html#hardware-required, "QEMU and Kubernetes” is broken.

Can you squash the commits?

intelkevinputnam commented:
> The structure looks good now. I noticed some dead links, for example in https://intelkevinputnam.github.io/pmem-csi/latest/install.html#hardware-required, "QEMU and Kubernetes” is broken.

Fixed, plus a few others. One of the reasons I like reStructuredText with Sphinx is the hardy internal reference system and automatic checking during the build.

> 
> Can you squash the commits?
Yep. Only took 2 tries.

pohly commented:
> Fixed, plus a few others. One of the reasons I like reStructuredText with Sphinx is the hardy internal reference system and automatic checking during the build.

So why did the broken link get through that checking?

> >    Can you squash the commits?
>
> Yep. Only took 2 tries.

Almost there :grin: Please also update the commit message of that resulting commit. You can do that with `git commit --amend`.

A good subject (first line) is "docs: generate doc site with Sphinx" followed by a blank line and then a short summary of what the commit is about. Usually we also expect a "why this is useful" or "why this way" but that's obvious in this case, so don't bother.

But I have some other comments:
- Can you move as many files as possible into the "docs" folder?
- "make html" is currently broken for me (see below).
- Can you add some instructions for developers to DEVELOPMENT.md? Consider what we need to know to continue maintaining this going forward in case that you are not around to help with it. For example, how do we deal with broken links? "bad-refs.txt  requirements.txt  substitutions.txt" all need to be documented.
- There a plenty of warnings during "make vhtml". Can we get rid of those? Ideally, warnings should be errors or suppressed.

Okay, so now we have "make html" and "make vhtml". The later is what Jenkins and I want to use, but it fails for me:
```
$ make vhtml
rm -rf _work/venv
python3 -m venv _work/venv
. _work/venv/bin/activate && pip install -r requirements.txt
Collecting sphinx (from -r requirements.txt (line 1))
...
. _work/venv/bin/activate && \
	sphinx-build -M vhtml "." "_output"   && \
	python3 fix-refs.py
Running Sphinx v3.0.2

Sphinx error:
Builder name vhtml not registered or available through entry point
make: *** [Makefile:220: vhtml] Error 2
```

Here's a fix:
```
diff --git a/Makefile b/Makefile
index 1c5f2fdd..a3c09fbb 100644
--- a/Makefile
+++ b/Makefile
@@ -218,12 +218,12 @@ BUILDDIR      = _output
 # Generate doc site under _build/html with Sphinx.
 vhtml: _work/venv/.stamp
        . _work/venv/bin/activate && \
-               $(SPHINXBUILD) -M $@ "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O) && \
+               $(SPHINXBUILD) -M html "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O) && \
                python3 fix-refs.py
                cp index.html $(BUILDDIR)/html/index.html
 
-html:
-               $(SPHINXBUILD) -M $@ "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O) && \
+html: 
+               $(SPHINXBUILD) -M html "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O) && \
                python3 fix-refs.py
                cp index.html $(BUILDDIR)/html/index.html
 ```

But why did it work in Jenkins? Answer: because it's not invoked in Jenkins anymore. I'm not sure yet why. Looking at the pipeline for https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/job/PR-562/16 I don't see the "docsite" stage. It's as if Jenkins ignored the modified Jenkinsfile. I wonder whether it does that intentionally because you are an untrusted contributor... those could cause all kinds of havoc if they were allowed to modify the pipeline. On the other hand, they can already do that by modifying other files, so this doesn't make a lot of sense.

pohly commented:
> It's as if Jenkins ignored the modified Jenkinsfile. I wonder whether it does that intentionally because you are an untrusted contributor...

That seems to be it. The docsite was produced (with my fix) in https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-571/3/pipeline.

intelkevinputnam commented:
@pohly Sphinx only checks cross referencing when they're written in reStructuredText and the syntax for those references are Sphinx specific, so they don't render well in GitHub. Markdown doesn't get the same royal treatment. I've added a couple of small extensions to improve markdown processing to fix references to files that need to be downloaded and references to other markdown files with anchors, but the checking is still a gap.

intelkevinputnam commented:
@pohly I've addressed most of your requests in another branch on my fork. Just working on the developer instructions for editing and improving the docs. You can take a look at the output [here](https://intelkevinputnam.github.io/pmem-csi/latest/README.html).

The build is down to one warning, and I was able to get rid of fix-refs.py and bad-refs.py. Those links are now fixed on the fly by changes I put in conf.py. You also now have the option to have linked files copied over into the html site or have the links converted to the github repo (by commit SHA when building with GitHub workflows). Links to directories are always turned into links to the repo. You can also choose to define a basebranch in the conf.py rather than use the commit SHA.

Once I'm done I'll squash and merge into this branch and amend the commit message.

intelkevinputnam commented:
@pohly Please, let me know if there is any more feedback.

pohly commented:
I ran `make vhtml` and still got one warning:
> WARNING: html_static_path entry '_static' does not exist

I indeed don't have `_static` directory. It looks like removing "html_static_path" from json.conf fixes this.

Commit b5e77867f35591588609a9236ac69a8e6c24c143 still has a rather long commit message:
```
docs: generate docs site with Sphinx
    
    Merge of squashed commits.
All of my changes for supporting docs and publishing with Sphinx.
    
    Signed-off-by: Kevin Putnam <kevin.putnam@intel.com>
    
    Removed local changes.
    
    Signed-off-by: Kevin Putnam <kevin.putnam@intel.com>
    
    Fixed typo in header.
...
```

As I want to try out this build with the modified Jenkinsfile, I'll create a separate PR myself where Jenkins will pick up that change and will remove that extra commit message comments there.

pohly commented:
Merged via PR #571, thanks @intelkevinputnam.
